### ISSUE TITLE
Add macOS to `local_offset_at`'s `OS_HAS_THREAD_SAFE_ENVIRONMENT`

### ISSUE THREAD

**Issue by PaulDance** [time-rs/time#718]
Dear maintainers,

I know this may seem like the 1298th duplicate issue about the underlying problem at first glance, but I have seen the previous issues regarding macOS *vs.* local offsets and still think something may have been missed. In any case, I'm open to discussion, hence my opening this.

Indeed, the `local_offset_at` has a few soundness checks, one of them being [`OS_HAS_THREAD_SAFE_ENVIRONMENT`](https://docs.rs/time/latest/src/time/sys/local_offset_at/unix.rs.html#12). When `true`, which is rarely the case, further soundness checks are disabled.

However, macOS is not part of the list, while what I'm seeing seems to indicate it should: although [FreeBSD's implementation](https://github.com/freebsd/freebsd-src/blob/4b65481ac68a99122991fd73583d071c5e27c65a/lib/libc/stdlib/getenv.c) is most probably not thread-safe and the [macOS man pages](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/getenv.3.html) do not mention it either, the actual Apple libc implementations of [`getenv`](https://github.com/apple-oss-distributions/Libc/blob/main/stdlib/FreeBSD/getenv.c#L118) and [`setenv`](https://github.com/apple-oss-distributions/Libc/blob/main/stdlib/FreeBSD/setenv.c#L446) really do seem to be thread-safe, which was added in [this commit](https://github.com/apple-oss-distributions/Libc/commit/337e89574a59b5259d3c953a425bbd3bb2ffff04#diff-552d1ab7bdb75937b99f59b0265fad4677cfc0facaa1f5e7dc1f000f2325866aR472), itself included as part of [macOS 10.12](https://github.com/apple-oss-distributions/distribution-macOS/tree/bee204afdfe063dfcb7921f8607b197d88417bb7).

Therefore, is there still a good reason not to include macOS in the thread-safe list? I've seen segmentation faults mentioned around here and specifically for this platform, but I would be surprised it would come from the environment manipulation itself.

Thanks in advance,
Paul.

**Comment by jhpratt** [comment]
No worries! It's immediately clear that this is not the same thing brought up repeatedly :)

Without digging into this again, you're 100% correct. I previously verified it, though after someone brought up that the behavior was relatively recent I removed the opt-out. While I don't recall the timeframe when this happened, it has almost certainly been long enough now.

With that said, there is no reason to open a pull request to add an opt-out here. On one of my SSDs I have a few commits that I haven't pushed yet, one of which removes the soundness checks entirely in favor of a better solution. These commits should be pushed in the coming days when I get a new laptop set up (hence the temporary storage).

**Comment by PaulDance** [comment]
Good, thanks for the quick answer, glad to hear!

**Comment by jhpratt** [comment]
This is on `main` and will be released shortly.

**Comment by PaulDance** [comment]
I saw febf3a10de60d7dd62223b03fbb05363bd0e6f90, and from what I understand, instead of calling `set_soundness`, one should now call `refresh_tz`, potentially as the first thing done in `main`, right? Also, I see that macOS was still not added to `OS_HAS_THREAD_SAFE_ENVIRONMENT` while it still could: was that intentional?

**Comment by jhpratt** [comment]
It was an oversight as I forgot to look back for verification. I was more focusing on the fact that the call will generally succeed out of the gate. I don't have a Mac to know if `refresh_tz` is needed at the start of the program, but on Linux (Fedora, specifically) I know that it is not.

cc @BlackHoleFox, who raised #610 originally. With [the minimum supported macOS increased](https://blog.rust-lang.org/2023/09/25/Increasing-Apple-Version-Requirements.html) last year, I think this is sufficient for soundness now.

**Comment by jhpratt** [comment]
So _right now_, I think not having it is correct because the previously mentioned change didn't occur until Rust 1.74. As the MSRV is 1.67.1, that's too soon. I have recently considered bumping the MSRV to something a fair bit higher (≥1.74) in the coming months, which would unblock this. Adding it to my tracker for that reason.
